### PR TITLE
Refactor validation functions to use Either instead of Validated

### PR DESCRIPTION
**Pull request by Alfhir**

### CODE REVIEW

File: src/main/kotlin/io/github/nomisrev/validation.kt [hunk 1]
```diff
@@ -1,15 +1,10 @@
-@file:Suppress("TooManyFunctions")
+@file:Suppress("TooManyFunctions", "WildcardImport")
```
Comment (nomisRev) [inline src/main/kotlin/io/github/nomisrev/validation.kt:1]: No wildcard imports please 😅 I never use them as a best practice.

File: src/main/kotlin/io/github/nomisrev/validation.kt [hunk 2]
```diff
@@ -51,88 +46,101 @@ data class InvalidEmailOrPassword(override val errors: NonEmptyList<String>) : I
...
 private const val MAX_USERNAME_LENGTH = 25
 
-private fun String.validPassword(): ValidatedNel<InvalidPassword, String> =
-  notBlank()
-    .zip(minSize(MIN_PASSWORD_LENGTH), maxSize(MAX_PASSWORD_LENGTH)) { a, _, _ -> a }
+private fun String.validPassword(): EitherNel<InvalidPassword, String> =
+  zipOrAccumulate(notBlank(), minSize(MIN_PASSWORD_LENGTH), maxSize(MAX_PASSWORD_LENGTH)) { a, _, _
+      ->
+      a
+    }
```
Comment (nomisRev) [inline src/main/kotlin/io/github/nomisrev/validation.kt:84]: Yikes, is this the formatting coming from spotless 😱 If yes, can we change this to:

```suggestion
  zipOrAccumulate(
    notBlank(),
    minSize(MIN_PASSWORD_LENGTH),
    maxSize(MAX_PASSWORD_LENGTH)
  ) { a, _, _ ->  a }
```

File: src/main/kotlin/io/github/nomisrev/validation.kt [hunk 3]
```diff
@@ -51,88 +46,101 @@ data class InvalidEmailOrPassword(override val errors: NonEmptyList<String>) : I
...
-  return trimmed
-    .notBlank()
-    .zip(trimmed.maxSize(MAX_EMAIL_LENGTH), trimmed.looksLikeEmail()) { a, _, _ -> a }
+  return zipOrAccumulate(
+      trimmed.notBlank(),
+      trimmed.maxSize(MAX_EMAIL_LENGTH),
+      trimmed.looksLikeEmail()
+    ) { a, _, _ ->
+      a
+    }
```
Comment (nomisRev) [inline src/main/kotlin/io/github/nomisrev/validation.kt:95]:
```suggestion
    ) { a, _, _ -> a }
```

File: src/main/kotlin/io/github/nomisrev/validation.kt [hunk 4]
```diff
@@ -51,88 +46,101 @@ data class InvalidEmailOrPassword(override val errors: NonEmptyList<String>) : I
...
-  return trimmed
-    .notBlank()
-    .zip(trimmed.minSize(MIN_USERNAME_LENGTH), trimmed.maxSize(MAX_USERNAME_LENGTH)) { a, _, _ ->
+  return zipOrAccumulate(
+      trimmed.notBlank(),
+      trimmed.minSize(MIN_USERNAME_LENGTH),
+      trimmed.maxSize(MAX_USERNAME_LENGTH)
+    ) { a, _, _ ->
       a
     }
```
Comment (nomisRev) [inline src/main/kotlin/io/github/nomisrev/validation.kt:107]:
```suggestion
    ) { a, _, _ -> a }
```

**Review by nomisRev — APPROVED** [review]
Looks perfect @Alfhir 🙌 🥳 
Only some nits about `*` import, and styling.

Thank you for the contribution, and happy hacktoberfest 🙌

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 2 above)
Comment (Alfhir) [reply src/main/kotlin/io/github/nomisrev/validation.kt:84]: Yes, and I think it's absolutely terrible, too. :-)
I think we will need to ignore spotless somehow, as the code looked like you propose before and spotless didn't let that pass. Or maybe override some rules somewhere? I never used spotless before, what would be your preferred way of solving this?

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 1 above)
Comment (Alfhir) [reply src/main/kotlin/io/github/nomisrev/validation.kt:1]: I'm fine with that. I think had already changed my settings to not use them, but then spotless handed me an error message I couldn't make sense of. Will try again.

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 2 above)
Comment (Alfhir) [reply src/main/kotlin/io/github/nomisrev/validation.kt:84]: `Caused by: org.gradle.api.GradleException: The following files had format violations:
    src/main/kotlin/io/github/nomisrev/validation.kt
        @@ -56,21 +56,21 @@
         
         fun·RegisterUser.validate():·Either<IncorrectInput,·RegisterUser>·=
         ··zipOrAccumulate(
        -····username.validUsername(),
        -····email.validEmail(),
        -····password.validPassword(),
        -····::RegisterUser
        -··)
        +······username.validUsername(),
        +······email.validEmail(),
        +······password.validPassword(),
        +······::RegisterUser
        +····)
         ····.mapLeft(::IncorrectInput)
         
         fun·Update.validate():·Either<IncorrectInput,·Update>·=
         ··zipOrAccumulate(
        -····username.mapOrAccumulate(String::validUsername),
        -····email.mapOrAccumulate(String::validEmail),
        -····password.mapOrAccumulate(String::validPassword)
        -··)·{·username,·email,·password·->
        -····Update(userId,·username,·email,·password,·bio,·image)
        -··}
        +······username.mapOrAccumulate(String::validUsername),
        +······email.mapOrAccumulate(String::validEmail),
        +······password.mapOrAccumulate(String::validPassword)
        +····)·{·username,·email,·password·->
        +······Update(userId,·username,·email,·password,·bio,·image)
        +····}
         ····.mapLeft(::IncorrectInput)
         
         private·fun·<E,·A,·B>·A?.mapOrAccumulate(f:·(A)·->·EitherNel<E,·B>):·EitherNel<E,·B?>·=
        @@ -83,30 +83,33 @@
         private·const·val·MAX_USERNAME_LENGTH·=·25
         
         private·fun·String.validPassword():·EitherNel<InvalidPassword,·String>·=
        -··zipOrAccumulate(
        -····notBlank(),
        -····minSize(MIN_PASSWORD_LENGTH),
        -····maxSize(MAX_PASSWORD_LENGTH)
        -··)·{·a,·_,·_·->·a·}
        +··zipOrAccumulate(notBlank(),·minSize(MIN_PASSWORD_LENGTH),·maxSize(MAX_PASSWORD_LENGTH))·{·a,·_,·_
        +······->
        +······a
        +····}
         ····.mapLeft(toInvalidField(::InvalidPassword))
         
         private·fun·String.validEmail():·EitherNel<InvalidEmail,·String>·{
    ... (31 more lines that didn't fit)`

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 1 above)
Comment (Alfhir) [reply src/main/kotlin/io/github/nomisrev/validation.kt:1]: Removed the wildcard imports.

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 2 above)
Comment (nomisRev) [reply src/main/kotlin/io/github/nomisrev/validation.kt:84]: I'll give it a try

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 1 above)
Comment (nomisRev) [reply src/main/kotlin/io/github/nomisrev/validation.kt:1]: Can you use those changes?

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 2 above)
Comment (nomisRev) [reply src/main/kotlin/io/github/nomisrev/validation.kt:84]: It's actually KtFmt, Spotless is just a Gradle wrapper, https://github.com/facebook/ktfmt
So not sure an issue at Spotless will help us, I checked and it's indeed forcing us that weird style.

Let's leave it for now, it seems we cannot opt a region out.

**Review by nomisRev — COMMENTED** [review]
Can you push the changes of removing the `*`?

File: src/main/kotlin/io/github/nomisrev/validation.kt (hunk 1 above)
Comment (Alfhir) [reply src/main/kotlin/io/github/nomisrev/validation.kt:1]: Thank you very much for getting back at me. I'll push the wildcard import changes this evening as soon as I have time.

**Comment by Alfhir** [comment]
Hi @nomisRev,

I'd like to learn a bit about OSS etiquette while doing this. So after I pushed changes, should I
- Do nothing (will you get notified?) or 
- Write a comment?

**Comment by nomisRev** [comment]
Hey @Alfhir,

> Do nothing (will you get notified?) or

I didn't get notified you pushed any changes 🤔 Even though I am subscribed to the PR. I think leaving a friendly reminder comment is never bad ☺️

**Review by nomisRev — APPROVED** [review]
These changes look perfect to me @Alfhir! Thank you 🙌 🥳 

Too about the horrible formatting of those lambdas..

**Comment by Alfhir** [comment]
> These changes look perfect to me @Alfhir! Thank you 🙌 🥳

Thank _you_ for providing the opportunity. It was great fun and if I can, I might do a few more of this to get to know Ktor better.

Have a great day

**Pull request merged**